source/common/stats/allocator_impl.cc

```diff
@@ -228,6 +228,25 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
   }
 };
 
+class TextReadoutImpl : public StatsSharedImpl<TextReadout> {
+public:
+  TextReadoutImpl(StatName name, AllocatorImpl& alloc, StatName tag_extracted_name,
+                  const StatNameTagVector& stat_name_tags)
+      : StatsSharedImpl(name, alloc, tag_extracted_name, stat_name_tags) {}
+
+  void removeFromSetLockHeld() EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) override {
+    const size_t count = alloc_.text_readouts_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void set(const std::string& value) override { value_ = value; }
+  std::string value() const override { return value_; }
+
+private:
+  std::string value_;
```

Comment:
quick drive-by: I think value_ needs to be atomic.

Do you want to have an 8-byte readout that uses an atomic<uint64_t>? Or mutex-protect the value?

Also you might be sharing some bits with counter/gauge that you don't need by inheriting from StatsSharedImpl -- can you check?

Reply:
If you are going to manage TextReadout via the Allocator, yes. The alternative is to treat them more like histograms, which has this disadvantage that if you create a histogram by the same name twice, you'll get two distinct objects. This results in some confusing output in the admin /stats panel.

So I'd go with ref-counting; you could just copy the code. I think the shared class between counters and gauges steals some bits leftover from flags() for the refcount, and you wouldn't be able to do that, but it's not really a big deal if you are paying for a std::string and mutex anyway.

Reply:
Thanks for the suggestion! 

I've tried to extract RefcountMetricImpl base class out of StatsSharedImpl, but in this case StatsSharedImpl only contains std::atomic<uint64_t> value_{0} which seems silly.

I've extracted value_ into Counter and Gauge instead and AFAICT there is nothing else in StatsSharedImpl that is unused by TextReadout.

[I've made the change locally, but having troubles with `git push origin textstats` claiming that "Everything up-to-date"]

Reply:
I'm not the best person to advise you on git-fu; I am barely alive on the platform :)

However I finally just pulled up the code and I think you can just move the value_ field out of StatsSharedImpl and just put it in both CounterImpl and GaugeImpl. value_ is not ref'd directly by anything in StatsSharedImpl anyway. You'll need flags_ because used() is probably still important.

Then you can use StatsSharedImpl from all 3 stat variants, and not worry about ref-counting.

Reply:
Perfect, that's exactly what I did. :) I'll ping you back once I manage to upload my changes.

Reply:
I've managed to push my changes, which is good, but somehow picked up unrelated commits, which seems bad. 

I'll be glad to fix that if you have any suggestions or try to create another PR from a clean branch.

Reply:
It looks like I've been able to get rid of unrelated commits, so PTAL when you have a chance.